Sources/CoreMetrics/Metrics.swift

```diff
@@ -488,6 +580,76 @@ public protocol MetricsFactory {
     func destroyTimer(_ handler: TimerHandler)
 }
 
+/// Wraps a CounterHandler, adding support for incrementing by floating point values by storing an accumulated floating point value and recoding increments to the underlying CounterHandler after crossing integer boundaries.
+private class AccumulatingRoundingFloatingPointCounter: FloatingPointCounterHandler {
+    private let lock = Lock()
+    private var fraction: Double = 0
+    private let counterHandler: CounterHandler
+
+    init(counterHandler: CounterHandler) {
+        self.counterHandler = counterHandler
+    }
+
+    func increment(by amount: Double) {
+        // FIXME: what do?
+        let maxIncrement = Double(Int64.max - (2 << 8))
+        assert((0...maxIncrement).contains(amount))
+        let amount = min(amount, maxIncrement)
+
+        // Split amount into integer and fraction components
+        var amountInteger = Int64(amount)
+        let amountFraction = amount.remainder(dividingBy: 1.0)
+        self.lock.withLockVoid {
+            // Add the fractional component to the accumulated fraction.
+            self.fraction += amountFraction
+            // After adding the fractional component of amount, the value may be
+            // larger than 1. Add any integer component of the accumulated
+            // fraction to the amount integer component and remove it from the
+            // accumulated fraction.
+            amountInteger += Int64(self.fraction)
+            self.fraction.formRemainder(dividingBy: 1.0)
+            // Increment the handler by the total integer component.
+            self.counterHandler.increment(by: amountInteger)
+        }
+    }
+
+    func reset() {
+        self.lock.withLockVoid {
+            self.fraction = 0
+            self.counterHandler.reset()
+        }
+    }
+
+    func destroy() {
+        MetricsSystem.factory.destroyCounter(self.counterHandler)
+    }
+}
+
+extension MetricsFactory {
+    /// Create a default backing `FloatingPointCounterHandler` for backends which do not naively support floating point counters.
+    ///
+    /// The created FloatingPointCounterHandler is a wrapper around a backend's CounterHandler which accumulates floating point increments and records increments to an underlying CounterHandler after crossing integer boundaries.
+    ///
+    /// - parameters:
+    ///     - label: The label for the `FloatingPointCounterHandler`.
+    ///     - dimensions: The dimensions for the `FloatingPointCounterHandler`.
+    public func makeFloatingPointCounter(label: String, dimensions: [(String, String)]) -> FloatingPointCounterHandler {
+        return AccumulatingRoundingFloatingPointCounter(
+            counterHandler: self.makeCounter(label: label, dimensions: dimensions))
```

Comment:
I guess we can also simplify this by moving the counter handler creation into the ctor of AccumulatingRoundingFloatingPointCounter

e.g.

```
 public func makeFloatingPointCounter(label: String, dimensions: [(String, String)]) -> FloatingPointCounterHandler {
  return AccumulatingRoundingFloatingPointCounter(label: label, dimensions: dimensions)
}

private class AccumulatingRoundingFloatingPointCounter: FloatingPointCounterHandler {
     private let counterHandler: CounterHandler

     init(label: String dimension: [String: String]) {
         self.counterHandler = MetricsSystem.factory.makeCounter(label: label, dimension: dimension)
     }
```